mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,707 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+import copy
+import threading
+from queue import Queue
+from typing import Any, Dict, List, Sequence, Tuple
+
+import chex
+import flax
+import hydra
+import jax
+import jax.debug
+import jax.numpy as jnp
+import numpy as np
+import optax
+from colorama import Fore, Style
+from flax.core.frozen_dict import FrozenDict
+from omegaconf import DictConfig, OmegaConf
+from optax._src.base import OptState
+from rich.pretty import pprint
+
+from mava.evaluator import get_sebulba_eval_fn as get_eval_fn
+from mava.evaluator import make_ff_eval_act_fn
+from mava.networks import FeedForwardActor as Actor
+from mava.networks import FeedForwardValueNet as Critic
+from mava.systems.ppo.types import LearnerState, OptStates, Params, PPOTransition
+from mava.types import (
+    ActorApply,
+    CriticApply,
+    ExperimentOutput,
+    Observation,
+    SebulbaLearnerFn,
+)
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax_utils import merge_leading_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.sebulba_utils import ParamsSource, Pipeline, RecordTimeTo, ThreadLifetime
+from mava.utils.total_timestep_checker import sebulba_check_total_timesteps
+from mava.utils.training import make_learning_rate
+from mava.wrappers.episode_metrics import get_final_step_metrics
+
+
+def rollout(
+    key: chex.PRNGKey,
+    config: DictConfig,
+    rollout_pipeline: Pipeline,
+    params_source: ParamsSource,
+    apply_fns: Tuple,
```

Comment:
Gonna need to import these from types

```suggestion
    apply_fns: Tuple[ActorApplyFn, CriticApplyFn],
```